File: Dynamic Programming/0486.Predict the Winner/Solution.java
class Solution {
    public boolean PredictTheWinner(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n + 2][n + 2];
        for (int len = 1; len <= n; len++) {
            for (int l = 1; l + len - 1 <= n; l++) {
                int r = l + len - 1;
                int a = nums[l - 1] - dp[l + 1][r];
                int b = nums[r - 1] - dp[l][r - 1];
                dp[l][r] = Math.max(a, b);
            }
        }
        return dp[1][n] >= 0;
    }
}

File: Dynamic Programming/0486.Predict the Winner/Solution.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [[0 for _ in range(n + 3)] for _ in range(n + 3)]
        for m in range(1, n + 1):
            for l in range(1, n - m + 2):
                r = l + m - 1;
                a = nums[l - 1] - dp[l + 1][r]
                b = nums[r - 1] - dp[l][r - 1]
                dp[l][r] = max(a, b)
        return dp[1][n] >= 0